src/main/java/io/javalin/plugin/openapi/dsl/extractDocumentation.kt

```diff
@@ -14,34 +16,37 @@ import io.javalin.core.util.isKotlinMethodReference
 import io.javalin.core.util.lambdaField
 import io.javalin.core.util.methodReferenceReflectionMethodName
 import io.javalin.core.util.methodsNotDeclaredByObject
+import io.javalin.core.util.parentFieldReflectionMethodName
 import io.javalin.plugin.openapi.CreateSchemaOptions
 import io.javalin.plugin.openapi.annotations.OpenApi
 import io.javalin.plugin.openapi.annotations.asOpenApiDocumentation
 import io.javalin.plugin.openapi.annotations.pathInfo
 import io.javalin.plugin.openapi.annotations.scanForAnnotations
 import io.javalin.plugin.openapi.annotations.warnUserAboutPotentialBugs
+import io.javalin.plugin.openapi.handler.DocumentedHandler
+import io.javalin.plugin.openapi.handler.ParameterHandler
+import io.javalin.plugin.openapi.handler.functional.SerializableFunction
 import java.lang.reflect.Field
 import java.lang.reflect.Method
 import java.util.logging.Logger
-import kotlin.reflect.KFunction
+import kotlin.jvm.internal.FunctionReference
 import kotlin.reflect.jvm.javaMethod
 
 fun HandlerMetaInfo.extractDocumentation(options: CreateSchemaOptions): OpenApiDocumentation {
     val documentation = document()
 
     options.default?.apply(documentation)
 
-    val userDocumentation = if (handler is DocumentedHandler) {
-        handler.documentation
-    } else {
-        val openApiAnnotation: OpenApi? = getOpenApiAnnotationFromReference() ?: getOpenApiAnnotationFromHandler()
-        openApiAnnotation?.asOpenApiDocumentation()
-                ?: extractDocumentationWithPathScanning(options)
-                ?: document()
+    when (handler) {
+        is DocumentedHandler -> listOf(handler.documentation).forEach(documentation::apply)
+        is ParameterHandler<*> -> documentation.apply(this.copy(handler = handler.reference).extractDocumentation(options))
+        else -> getOpenApiAnnotationFromReference()
```

Comment:
Every handler is now either a `DocumentedHandler `, documented by dsl, a "pointer" `ParameterHandler` or a reflectable/unreflectable reference